### ISSUE TITLE
[ML] GET large trained models with definition errors

### ISSUE THREAD

**Issue by davidkyle** [elastic/elasticsearch#80254]
On a node with 1GB heap after uploading a large PyTorch model a GET with model definition errors out:

`GET _ml/trained_models/LARGE_PYTORCH_MODEL?include=definition`

This caused the node to stop without logging any errors but the suspicion is this query is to blame:

https://github.com/elastic/elasticsearch/blob/12ad399c488f0cc60e19b5e1b29c6d569cb4351a/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java#L615

`MAX_NUM_DEFINITION_DOCS` == 100 which is too large for retrieval given the definition docs are 4MB or 16MB (DFA models) in size.

It is not feasible to return the entire PyTorch model definition in one request, protection should be in place to prevent this and return an error if the model is a PyTorch model.

As DFA models get larger moving towards 100MB an API to export the model in chunks will be required.

**Comment by davidkyle** [comment]
A 4GB node fares better managing to return a response

```
{
	"error": {
		"root_cause": [{
			"type": "exception",
			"reason": "Model definition truncated. Unable to deserialize trained model definition [bert-model-fine-tuned-for-ner]"
		}],
		"type": "exception",
		"reason": "Model definition truncated. Unable to deserialize trained model definition [bert-model-fine-tuned-for-ner]"
	}
}
```

**Comment by benwtrent** [comment]
I think `include=definition` should just not work at all for pytorch models. What say you?

**Comment by davidkyle** [comment]
> I think include=definition should just not work at all for pytorch models. What say you?

totes

But I'm also concerned a large DFA model could cause the same problem on a node with a small heap and I'd like to revisit export of those models.

**Comment by benwtrent** [comment]
> But I'm also concerned a large DFA model could cause the same problem on a node with a small heap and I'd like to revisit export of those models.

Agreed. Lets get this fixed for v8 (no pytorch when including definitions) and do the boosted tree model fix in a separate PR.

**Comment by benwtrent** [comment]
I think the fix for boosted trees is somehow adding the parsed model (if definition is to be uncompressed) to the circuit breaker and releasing the bytes when we notify the listener. 

I am not 100% sure how to do this, but it should be possible

**Comment by benwtrent** [comment]
While investigating this issue I ran into: https://github.com/elastic/elasticsearch/pull/80377

Still not 100% sure how to prevent major errors when parsing XContent that is WAY too large. It may be we need to tell the circuit breaker we are going to use 2x the model size when returning to make sure we have headroom. Experimenting.

**Comment by benwtrent** [comment]
I think issues stem from:

 - Copying byte buffers around instead of using `ReleasableBytesReference readReleasableBytesReference()` for these large definitions (unsure how easy it will be to switch to that)
 - Holding the compressed version in memory WHILE parsing it for XContent response (seems like we could use less memory if we fully parsed it; if requested, right away). But this has issues as we don't serialize that parsed definition between nodes. Yet another good reason for using `ReleasableBytesReference`

Overall, we need a better way to retrieve large definitions.
